neural_lam/datastore/base.py

```diff
@@ -0,0 +1,371 @@
+        stacked. Where the value is 1, the grid point is a boundary point, and
+        where the value is 0, the grid point is not a boundary point.
+
+        Returns
+        -------
+        xr.DataArray
+            The boundary mask for the dataset, with dimensions
+            `('grid_index',)`.
+
+        """
+        pass
+
+
+@dataclasses.dataclass
+class CartesianGridShape:
+    """Dataclass to store the shape of a grid."""
+
+    x: int
+    y: int
+
+
+class BaseCartesianDatastore(BaseDatastore):
+    """Base class for weather
+    data stored on a Cartesian
+    grid. In addition to the
+    methods and attributes
+    required for weather data
+    in general (see
+    `BaseDatastore`) for
+    Cartesian gridded source
+    data each `grid_index`
+    coordinate value is assume
+    to have an associated `x`
+    and `y`-value so that the
+    processed data-arrays can
+    be reshaped back into into
+    2D xy-gridded arrays.
+
+    In addition the following attributes and methods are required:
+    - `coords_projection` (property): Projection object for the coordinates.
+    - `grid_shape_state` (property): Shape of the grid for the state variables.
+    - `get_xy_extent` (method): Return the extent of the x, y coordinates for a
+      given category of data.
+    - `get_xy` (method): Return the x, y coordinates of the dataset.
+
+    """
+
+    CARTESIAN_COORDS = ["y", "x"]
+
+    @property
+    @abc.abstractmethod
+    def coords_projection(self) -> ccrs.Projection:
+        """Return the projection object for the coordinates.
+
+        The projection object is used to plot the coordinates on a map.
+
+        Returns
+        -------
+        cartopy.crs.Projection:
+            The projection object.
+
+        """
+        pass
+
+    @property
+    @abc.abstractmethod
+    def grid_shape_state(self) -> CartesianGridShape:
+        """The shape of the grid for the state variables.
+
+        Returns
+        -------
+        CartesianGridShape:
+            The shape of the grid for the state variables, which has `x` and
+            `y` attributes.
+
+        """
+        pass
+
+    @abc.abstractmethod
+    def get_xy(self, category: str, stacked: bool) -> np.ndarray:
+        """Return the x, y coordinates of the dataset.
+
+        Parameters
+        ----------
+        category : str
+            The category of the dataset (state/forcing/static).
+        stacked : bool
+            Whether to stack the x, y coordinates.
+
+        Returns
+        -------
+        np.ndarray
+            The x, y coordinates of the dataset, returned differently based on
+            the value of `stacked`:
+            - `stacked==True`: shape `(2, n_grid_points)` where
+                               n_grid_points=N_x*N_y.
+            - `stacked==False`: shape `(2, N_y, N_x)`
+
+        """
+        pass
+
+    def get_xy_extent(self, category: str) -> List[float]:
```

Comment:
Propose to add a cache decorator to this: https://docs.python.org/dev/library/functools.html#functools.lru_cache so it does not have to be-recomputed for each plot. Alternatively pre-compute for each category in constructor.

Reply:
True, it would be a good idea to cache the result of this function, although the decorator would need to be added to implementations of this class not the abstract base class.

Reply:
But this function isn't abstract here, it actually implements the computation of the extent?

Reply:
OH! Good point sorry, I missread the where the comment was linked to. I will add this

Reply:
I added this caching to `get_xy_extent` in https://github.com/leifdenby/neural-lam/commit/ac10d7df3c25d7fd15437363eecc5d839f484b5b